framework/include/lightsource_gpu_data.hpp

```diff
@@ -0,0 +1,67 @@
+#pragma once
+
+namespace cgb
+{
+	/** Light data in the right format to be uploaded to the GPU
+	* and to be used in a GPU buffer like a UBO or an SSBO
+	*/
+	struct lightsource_gpu_data
```

Comment:
I would not consider the memory layout of this type optimal. It can not be used with the `std140` memory layout, for example, which requires each element to be padded to 16 bytes (i.e. `vec4`). See [OpenGL Wiki Interface Block (GLSL)](https://www.khronos.org/opengl/wiki/Interface_Block_(GLSL)#Memory_layout) for details. It works with `std430`, but I would prefer not to have such a restriction. Furthermore, alignment to 16-byte boundaries should result in more efficient memory access.

Therefore, the following struct definition is suggested:
``` 
	struct lightsource_gpu_data
	{
		/** Ambient color of the light source. */
		alignas(16) glm::vec4 mColorAmbient;
		/** Diffuse color of the light source. */
		alignas(16) glm::vec4 mColorDiffuse;
		/** Specular color of the light source. */
		alignas(16) glm::vec4 mColorSpecular;
		/** Direction of the light source. */
		alignas(16) glm::vec4 mDirection;
		/** Position of the light source. */
		alignas(16) glm::vec4 mPosition;
		/** Angles, where the individual elements contain the following data:
		 *  [0] ... inner cone angle
		 *  [1] ... outer cone angle
		 *  [2] ... unused
		 *  [3] ... unused
		 */
		alignas(16) glm::vec4 mAngles;
		/** Light source attenuation, where the individual elements contain the following data:
		 *  [0] ... constant attenuation factor
		 *  [1] ... linear attenuation factor
		 *  [2] ... quadratic attenuation factor
		 *  [3] ... unused
		 */
		alignas(16) glm::vec4 mAttenuation;

		/** General information about the light source, where the individual elements contain the following data:
		 *  [0] ... type of the light source
		 */
		alignas(16) glm::ivec4 mInfo;
	};
``` 

Using only `glm::vec4` and `glm::ivec4` means that the `alignas` specifiers would not be required anymore, but it doesn't hurt to leave them in place.